doc/tutorials/inter-vm_communication.rst

```diff
@@ -0,0 +1,490 @@
+        {
+            /* unmap reg mmio space from BAR0 */
+            munmap(p_ivsh_dev_ctx->p_reg, IVSH_BAR0_SIZE);
+            p_ivsh_dev_ctx->p_reg = NULL;
+            close(p_ivsh_dev_ctx->bar0_fd);
+            p_ivsh_dev_ctx->bar0_fd = -1;
+
+            /* unmap shared memory from BAR2 */
+            munmap(p_ivsh_dev_ctx->p_shmem, p_ivsh_dev_ctx->shmem_size);
+            p_ivsh_dev_ctx->p_shmem = NULL;
+            close(p_ivsh_dev_ctx->bar2_fd);
+            p_ivsh_dev_ctx->bar2_fd = -1;
+
+            /* used for doorbell mode*/
+            for i←0 to IVSH_MAX_IRQ_NUM
+                do
+                    close(p_ivsh_dev_ctx->irq_data[i].fd);
+                    p_ivsh_dev_ctx->irq_data[i].fd = -1;
+                    close(p_ivsh_dev_ctx->epfds_irq[i]);
+                    p_ivsh_dev_ctx->epfds_irq[i] = -1;
+                end
+            close(p_ivsh_dev_ctx->uio_dev_fd);
+            p_ivsh_dev_ctx->uio_dev_fd = -1;
+        }
+
+**The following reference code are used for Inter-VM communication based on Doorbell mode:**
+
+- Trigger ivshmem doorbell
+
+     .. code-block:: none
+
+        void ivsh_trigger_doorbell(struct ivsh_dev_context *p_ivsh_dev_ctx, uint16_t peer_id, uint16_t vector_id)
+        {
+            p_ivsh_dev_ctx->p_reg[IVSH_REG_DOORBELL >> 2] = (peer_id << 16) | vector_id;
+        }
+
+
+- Wait ivshmem device irq
+
+     .. code-block:: none
+
+        static inline int ivsh_wait_irq(struct ivsh_dev_context *p_ivsh_dev_ctx, unsigned int idx)
+        {
+            struct epoll_event ev = {0};
+            struct uio_irq_data *irq_data = NULL;
+            eventfd_t val;
+            int n;
+            while (1) {
+                n = epoll_wait(p_ivsh_dev_ctx->epfds_irq[idx], &ev, 1, -1);
+                if (n == 1) {
+                    irq_data = ev.data.ptr;
+                    eventfd_read(irq_data->fd, &val);
+                    break;
+                }
+            }
+        }
+
+**Data Transfer State-Machine**
+
+  A state machine is introduced as a communication mechanism between the two VMs,
+  which use the same ivshmem PCI device for the data transfer.
+
+  It includes three states – RESET, READY, and INIT. RESET state is the initial state
+  after ivshmem device is initialized.
+
+    - When both VM states are in the RESET, Sender VM prepares the sending data
+      and then sets its state to INIT, and Receiver VM prepares receiving buffer
+      and sets its state to INIT.
+
+    - When both VM's state is in the INIT, Sender VM sets its state in the READY
+      after sending all of the data, and Receiver VM sets its state in the READY
+      after receiving all of the transfer data.
+
+    - Then both VMs change their status from READY to INIT for starting the next
+      round of data transfer.
+
+  :numref:`Inter-VM ivshmem data transfer state machine` shows the state machine relationship:
+
+    .. figure:: images/Inter-VM_data_transfer_state_machine.png
```

Comment:
Watch the indenting of all this figure directive and the sentence above.  If it's not part of the bullet list, it should not be indented at all.

Reply:
Thanks, has been fixed.